flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/planner/Planner.java

```diff
@@ -0,0 +1,101 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.streaming.api.transformations.StreamTransformation;
+import org.apache.flink.table.api.QueryConfig;
+import org.apache.flink.table.operations.ModifyTableOperation;
+import org.apache.flink.table.operations.QueryTableOperation;
+import org.apache.flink.table.operations.TableOperation;
+
+import java.util.List;
+
+/**
+ * This interface serves two purposes:
+ * <ul>
+ * <li>SQL parser - transforms a SQL string into a Table API specific tree of
+ * {@link TableOperation}s</li>
+ * <li>relational planner - provides a way to plan, optimize and transform tree of
+ * {@link ModifyTableOperation} into a runnable form ({@link StreamTransformation})</li>
+ * </ul>.
+ *
+ * <p>The Planner is execution agnostic. It is up to the
+ * {@link org.apache.flink.table.api.TableEnvironment} to ensure that if any of the
+ * {@link QueryTableOperation} pull any runtime configuration, all those configurations are
+ * equivalent. Example: If some of the {@link QueryTableOperation}s scan DataStreams, all
+ * those DataStreams must come from the same StreamExecutionEnvironment, because the result
+ * of {@link Planner#translate(List, QueryConfig)} will strip any execution configuration from
+ * the DataStream information.
+ *
+ * <p>All Tables referenced in either {@link Planner#parse(String)} or
+ * {@link Planner#translate(List, QueryConfig)} should be previously registered in a
+ * {@link org.apache.flink.table.catalog.CatalogManager}, which will be provided during
+ * instantiation of the {@link Planner}.
+ */
+@Internal
+public interface Planner {
+
+	/**
+	 * Entry point for parsing sql queries expressed as a String.
+	 *
+	 * <p><b>Note:</b>If the created {@link TableOperation} is a {@link QueryTableOperation}
+	 * it must be in a form that will be understood by the
+	 * {@link Planner#translate(List, QueryConfig)} method.
+	 *
+	 * <p>The produced Operation tree should already be validated.
+	 *
+	 * @param stmt the sql statement to evaluate
+	 * @return parsed query as a tree of relational {@link TableOperation}s
+	 */
+	TableOperation parse(String stmt);
+
+	/**
+	 * Converts a relational tree of {@link QueryTableOperation}s into a set of runnable
+	 * {@link StreamTransformation}s.
+	 *
+	 * <p>This method accepts a list of {@link QueryTableOperation}s to allow reusing common
+	 * subtrees of multiple relational queries.
+	 *
+	 * @param tableOperations list of relational operations to plan, optimize and convert in a
+	 * single run.
+	 * @param queryConfig allows configuring runtime behavior of the resulting query
+	 * @return list of corresponding {@link StreamTransformation}s.
+	 */
+	List<StreamTransformation<?>> translate(List<ModifyTableOperation> tableOperations, QueryConfig queryConfig);
```

Comment:
`QueryConfig` looses its purpose with a unified table environment that might merge parts of a table operation tree into a common part. We should merge query config into table config. Table config can be considered a mutable session configuration that can be updated at any time. It can be passed to the constructor of a planner. Thus, we can remove the `QueryConfig` parameter from all methods. What do you think?